File: src/kicad_xyce_plugin/simulation_parameters/op_simulation_parameters.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from ..netlist_parser import NetlistTopology

from .print_parameters import PrintParameters


@dataclass(frozen=True)
class NodesetEntry:
    node: str
    voltage: str


@dataclass(frozen=True)
class IcEntry:
    node: str
    voltage: str


@dataclass(frozen=True)
class OpSimulationParameters:

    print_dc_enabled: bool = False
    print_dc_all_nodes: bool = False
    print_dc_all_currents: bool = False
    print_dc_specific_variables: tuple[str, ...] = ()
    print_dc_format: str = ""
    print_dc_file: str = ""
    save_enabled: bool = False
    save_type: str = "NODESET"
    save_file: str = ""
    nodeset_entries: tuple[NodesetEntry, ...] = ()
    ic_entries: tuple[IcEntry, ...] = ()
    replace_ground: bool = True
    print_parameters: PrintParameters | None = None

    @classmethod
    def from_xyce_directives(cls, directives: list[str]) -> "OpSimulationParameters" | None:
        # preprocess replaceground flag
        replace_ground = True
        # init flag
        print_dc_enabled = False
        # init list
        print_dc_vars = []
        # init parsed print parameters object
        print_parameters_parsed: PrintParameters | None = None
        # init flag
        save_enabled = False
        # init list
        nodeset_entries = []
        # init list
        ic_entries = []
        # flag indicating whether a valid directive was found
        found = False
        # parse directives
        for directive in directives:
            # get tokens
            tokens = directive.split()
            # check command
            cmd = tokens[0].upper()
            # check .OP directive
            if cmd == ".OP":
                # set flag
                found = True
                # next
                continue
            # handle print dc
            if cmd == ".PRINT" and len(tokens) > 1 and tokens[1].upper() == "DC":
                # set enabled
                print_dc_enabled = True
                # parse via PrintParameters for structured wildcard/format access
                print_parameters_parsed = PrintParameters.from_xyce_statement(directive)
                # also collect raw vars for backward-compatible list
                print_dc_vars.extend(tokens[2:])
                # next
                continue
            # handle save
            if cmd == ".SAVE":
                # set enabled
                save_enabled = True
                # next
                continue
            # handle nodeset
            if cmd == ".NODESET":
                # process pairs
                for pair in tokens[1:]:
                    # check if pair valid
                    if "=" in pair:
                        # split node and voltage
                        node_part, voltage = pair.split("=", 1)
                        # validate
                        if node_part.startswith("V(") and node_part.endswith(")"):
                            # append entry
                            nodeset_entries.append(NodesetEntry(node=node_part[2:-1], voltage=voltage))
                # next
                continue
            # handle initial conditions (.IC and .DCVOLT use the same format)
            if cmd in (".IC", ".DCVOLT"):
                # iterate tokens looking for V(node)=val or node val pairs
                remaining = tokens[1:]
                i = 0
                while i < len(remaining):
                    token = remaining[i]
                    if "=" in token:
                        # V(node)=val form
                        lhs, voltage = token.split("=", 1)
                        if lhs.startswith("V(") and lhs.endswith(")"):
                            node = lhs[2:-1]
                        else:
                            node = lhs
                        ic_entries.append(IcEntry(node=node, voltage=voltage))
                        i += 1
                    elif i + 1 < len(remaining):
                        # node val pair form
                        ic_entries.append(IcEntry(node=token, voltage=remaining[i + 1]))
                        i += 2
                    else:
                        i += 1
                # next
                continue
            # preprocess replaceground
            if cmd == ".PREPROCESS" and len(tokens) > 2 and tokens[1].upper() == "REPLACEGROUND":
                # set flag based on value
                replace_ground = tokens[2].upper() == "TRUE"
        # return instance if a valid directive was found
        return cls(print_dc_enabled=print_dc_enabled, print_dc_specific_variables=tuple(print_dc_vars), save_enabled=save_enabled, nodeset_entries=tuple(nodeset_entries), ic_entries=tuple(ic_entries), replace_ground=replace_ground, print_parameters=print_parameters_parsed) if found else None

    def to_xyce_directives(self, topology: NetlistTopology | None = None) -> list[str]:
        # start lines
        lines = [".OP"]
        # use print_parameters when set (newer wildcard approach)
        if self.print_parameters is not None:
            # emit the statement directly using the structured print parameters
            lines.append(self.print_parameters.to_xyce_statement())
        # check enabled
        elif self.print_dc_enabled:
            # start tokens
            tokens = [".PRINT DC"]
            # check format
            if self.print_dc_format:
                # append format
                tokens.append(f"FORMAT={self.print_dc_format}")
            # check file
            if self.print_dc_file:
                # append file
                tokens.append(f"FILE={self.print_dc_file}")
            # get custom vars
            vars = list(self.print_dc_specific_variables)
            # check topology
            if topology:
                # check all nodes
                if self.print_dc_all_nodes:
                    # iterate nodes
                    for node in topology.nodes:
                        # append node
                        vars.append(f"V({node})")
                # check all currents
                if self.print_dc_all_currents:
                    # iterate devices
                    for dev in topology.devices:
                        # append current
                        vars.append(f"I({dev.name})")
            # add unique
            tokens.extend(dict.fromkeys(vars))
            # join tokens
            lines.append(" ".join(tokens))
        # check save enabled
        if self.save_enabled:
            # build tokens
            tokens = [".SAVE"]
            # append type
            tokens.append(f"TYPE={self.save_type}")
            # check file
            if self.save_file:
                # append file
                tokens.append(f"FILE={self.save_file}")
            # join tokens
            lines.append(" ".join(tokens))
        # check nodeset entries
        if self.nodeset_entries:
            # format pairs
            pairs = " ".join(f"V({e.node})={e.voltage}" for e in self.nodeset_entries)
            # append directive
            lines.append(f".NODESET {pairs}")
        # check initial condition entries
        if self.ic_entries:
            # format pairs
            pairs = " ".join(f"V({e.node})={e.voltage}" for e in self.ic_entries)
            # append directive
            lines.append(f".IC {pairs}")
        # return directives
        return ([".PREPROCESS REPLACEGROUND TRUE"] if self.replace_ground else []) + lines

    def raw_output_file_path(self, working_directory: Path, netlist_file_path: Path) -> Path | None:
        # check raw format is selected in print parameters and a print file is specified
        if self.print_parameters is None or self.print_parameters.print_format != "RAW":
            return None
        # return the output file path specified in the print parameters if available
        if self.print_parameters.print_file:
            return working_directory / self.print_parameters.print_file
        # otherwise, create raw file from netlist file, adding the ".raw" suffix
        return netlist_file_path.with_suffix(netlist_file_path.suffix + ".raw")

File: src/simulation_parameters/op_simulation_parameters.cpp
#include <algorithm>
#include <cctype>
#include <optional>
#include <string>
#include <vector>

#include "op_simulation_parameters.h"

// normalize a string to uppercase
static std::string op_to_upper(std::string s) {
    // convert each character to upper case
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::toupper(c); });
    // return converted string
    return s;
}

// tokenize a directive by whitespace
static std::vector<std::string> op_tokenize(const std::string& directive) {
    // init token list
    std::vector<std::string> tokens;
    // init current token buffer
    std::string current;
    // iterate characters
    for (const char ch : directive) {
        // check whitespace splitter
        if (std::isspace(static_cast<unsigned char>(ch))) {
            // flush current token when non-empty
            if (!current.empty()) {
                tokens.push_back(current);
                current.clear();
            }
            // next
            continue;
        }
        // append char
        current += ch;
    }
    // flush trailing token
    if (!current.empty()) {
        tokens.push_back(current);
    }
    // return tokens
    return tokens;
}

NodesetEntry::NodesetEntry(std::string node, std::string voltage) :
    node(std::move(node)), voltage(std::move(voltage)) {}

bool NodesetEntry::operator==(const NodesetEntry& other) const { return node == other.node && voltage == other.voltage; }

IcEntry::IcEntry(std::string node, std::string voltage) :
    node(std::move(node)), voltage(std::move(voltage)) {}

bool IcEntry::operator==(const IcEntry& other) const { return node == other.node && voltage == other.voltage; }

OpSimulationParameters::OpSimulationParameters(bool print_dc_enabled, bool print_dc_all_nodes, bool print_dc_all_currents, std::vector<std::string> print_dc_specific_variables, std::string print_dc_format, std::string print_dc_file, bool save_enabled, std::string save_type, std::string save_file, std::vector<NodesetEntry> nodeset_entries, std::vector<IcEntry> ic_entries, bool replace_ground, std::optional<PrintParameters> print_parameters) :
    print_dc_enabled(print_dc_enabled), print_dc_all_nodes(print_dc_all_nodes), print_dc_all_currents(print_dc_all_currents), print_dc_specific_variables(std::move(print_dc_specific_variables)), print_dc_format(std::move(print_dc_format)), print_dc_file(std::move(print_dc_file)), save_enabled(save_enabled), save_type(std::move(save_type)), save_file(std::move(save_file)), nodeset_entries(std::move(nodeset_entries)), ic_entries(std::move(ic_entries)), replace_ground(replace_ground), print_parameters(std::move(print_parameters)) {}

std::optional<OpSimulationParameters> OpSimulationParameters::from_xyce_directives(const std::vector<std::string>& directives) {
    // preprocess replaceground flag
    bool replace_ground = true;
    // init flag
    bool print_dc_enabled = false;
    // init list
    std::vector<std::string> print_dc_vars;
    // init parsed print parameters object
    std::optional<PrintParameters> print_parameters_parsed;
    // init flag
    bool save_enabled = false;
    // init list
    std::vector<NodesetEntry> nodeset_entries;
    // init list
    std::vector<IcEntry> ic_entries;

    // flag indicating whether a valid directive was found
    bool found = false;

    // parse directives
    for (const auto& directive : directives) {
        // tokenize the directive
        const auto tokens = op_tokenize(directive);

        // skip empty directives
        if (tokens.empty()) {
            continue;
        }

        const std::string cmd = op_to_upper(tokens[0]);

        // check .OP directive
        if (cmd == ".OP") {
            // set flag
            found = true;
            continue;
        }

        // handle print dc
        if (cmd == ".PRINT" && tokens.size() > 1 && op_to_upper(tokens[1]) == "DC") {
            // set enabled
            print_dc_enabled = true;
            // parse via PrintParameters for structured wildcard/format access
            const auto print_parameters_parsed_opt = PrintParameters::from_xyce_statement(directive);
            if (print_parameters_parsed_opt) {
                print_parameters_parsed = *print_parameters_parsed_opt;
            }
            // also collect raw vars for backward-compatible list
            for (size_t i = 2; i < tokens.size(); ++i) {
                print_dc_vars.push_back(tokens[i]);
            }
            continue;
        }

        // handle save
        if (cmd == ".SAVE") {
            // set enabled
            save_enabled = true;
            continue;
        }

        // handle nodeset
        if (cmd == ".NODESET") {
            // process pairs
            for (size_t i = 1; i < tokens.size(); ++i) {
                const auto& pair = tokens[i];
                // check if pair valid
                if (pair.find('=') != std::string::npos) {
                    // split node and voltage
                    const auto eq_pos = pair.find('=');
                    const std::string node_part = pair.substr(0, eq_pos);
                    const std::string voltage = pair.substr(eq_pos + 1);
                    // validate
                    if (node_part.substr(0, 2) == "V(" && node_part.back() == ')') {
                        // append entry
                        nodeset_entries.emplace_back(node_part.substr(2, node_part.length() - 3), voltage);
                    }
                }
            }
            continue;
        }

        // handle initial conditions (.IC and .DCVOLT use the same format)
        if (cmd == ".IC" || cmd == ".DCVOLT") {
            // iterate tokens looking for V(node)=val or node val pairs
            for (size_t i = 1; i < tokens.size(); ++i) {
                const auto& token = tokens[i];
                if (token.find('=') != std::string::npos) {
                    // V(node)=val form
                    const auto eq_pos = token.find('=');
                    const std::string lhs = token.substr(0, eq_pos);
                    const std::string voltage = token.substr(eq_pos + 1);
                    std::string node;
                    if (lhs.substr(0, 2) == "V(" && lhs.back() == ')') {
                        node = lhs.substr(2, lhs.length() - 3);
                    }
                    else {
                        node = lhs;
                    }
                    ic_entries.emplace_back(node, voltage);
                }
                else if (i + 1 < tokens.size()) {
                    // node val pair form
                    ic_entries.emplace_back(token, tokens[i + 1]);
                    ++i;
                }
            }
            continue;
        }

        // preprocess replaceground
        if (cmd == ".PREPROCESS" && tokens.size() > 2 && op_to_upper(tokens[1]) == "REPLACEGROUND") {
            // set flag based on value
            replace_ground = (op_to_upper(tokens[2]) == "TRUE");
        }
    }

    // return instance if a valid directive was found
    if (!found) {
        return std::nullopt;
    }

    return OpSimulationParameters(print_dc_enabled, false, false, print_dc_vars, "", "", save_enabled, "NODESET", "", nodeset_entries, ic_entries, replace_ground, print_parameters_parsed);
}

std::vector<std::string> OpSimulationParameters::to_xyce_directives() const {
    // init output directive list
    std::vector<std::string> directives;

    // start with the .OP directive
    directives.push_back(".OP");

    // use print_parameters when set (newer wildcard approach)
    if (print_parameters) {
        // emit the statement directly using the structured print parameters
        directives.push_back(print_parameters->to_xyce_statement());
    }
    // check enabled
    else if (print_dc_enabled) {
        // start tokens
        std::string tokens = ".PRINT DC";
        // check format
        if (!print_dc_format.empty()) {
            // append format
            tokens += " FORMAT=" + print_dc_format;
        }
        // check file
        if (!print_dc_file.empty()) {
            // append file
            tokens += " FILE=" + print_dc_file;
        }
        // add custom vars
        for (const auto& var : print_dc_specific_variables) {
            tokens += " " + var;
        }
        directives.push_back(tokens);
    }

    // check save enabled
    if (save_enabled) {
        // build tokens
        std::string tokens = ".SAVE";
        // append type
        tokens += " TYPE=" + save_type;
        // check file
        if (!save_file.empty()) {
            // append file
            tokens += " FILE=" + save_file;
        }
        directives.push_back(tokens);
    }

    // check nodeset entries
    if (!nodeset_entries.empty()) {
        // format pairs
        std::string pairs;
        for (const auto& entry : nodeset_entries) {
            if (!pairs.empty()) {
                pairs += " ";
            }
            pairs += "V(" + entry.node + ")=" + entry.voltage;
        }
        // append directive
        directives.push_back(".NODESET " + pairs);
    }

    // check initial condition entries
    if (!ic_entries.empty()) {
        // format pairs
        std::string pairs;
        for (const auto& entry : ic_entries) {
            if (!pairs.empty()) {
                pairs += " ";
            }
            pairs += "V(" + entry.node + ")=" + entry.voltage;
        }
        // append directive
        directives.push_back(".IC " + pairs);
    }

    // prepend replaceground preprocessing when enabled
    if (replace_ground) {
        directives.insert(directives.begin(), ".PREPROCESS REPLACEGROUND TRUE");
    }

    // return directives
    return directives;
}

bool OpSimulationParameters::operator==(const OpSimulationParameters& other) const { return print_dc_enabled == other.print_dc_enabled && print_dc_all_nodes == other.print_dc_all_nodes && print_dc_all_currents == other.print_dc_all_currents && print_dc_specific_variables == other.print_dc_specific_variables && print_dc_format == other.print_dc_format && print_dc_file == other.print_dc_file && save_enabled == other.save_enabled && save_type == other.save_type && save_file == other.save_file && nodeset_entries == other.nodeset_entries && ic_entries == other.ic_entries && replace_ground == other.replace_ground && print_parameters == other.print_parameters; }
